From ad2f0a1e65bd58dee78804dc28736ff616381356 Mon Sep 17 00:00:00 2001
From: Bo-Cun Chen <bc-bocun.chen@mediatek.com>
Date: Wed, 27 Nov 2024 10:16:27 +0800
Subject: [PATCH 78/84] net: ethernet: mtk_eth_soc: support forced reset
 control

Without this patch, users are unable to control the forced reset
through the proprietary debugfs.

Signed-off-by: Bo-Cun Chen <bc-bocun.chen@mediatek.com>
---
 drivers/net/ethernet/mediatek/mtk_eth_soc.c | 6 ++++--
 drivers/net/ethernet/mediatek/mtk_eth_soc.h | 1 +
 2 files changed, 5 insertions(+), 2 deletions(-)

diff --git a/drivers/net/ethernet/mediatek/mtk_eth_soc.c b/drivers/net/ethernet/mediatek/mtk_eth_soc.c
index 799dc58e2581..730146eb322e 100644
--- a/drivers/net/ethernet/mediatek/mtk_eth_soc.c
+++ b/drivers/net/ethernet/mediatek/mtk_eth_soc.c
@@ -3609,7 +3609,8 @@ static void mtk_tx_timeout(struct net_device *dev, unsigned int txqueue)
 	eth->netdev[mac->id]->stats.tx_errors++;
 	netif_err(eth, tx_err, dev, "transmit timed out\n");
 
-	schedule_work(&eth->pending_work);
+	if (atomic_read(&eth->reset.force))
+		schedule_work(&eth->pending_work);
 }
 
 static int mtk_get_irqs_fe(struct platform_device *pdev, struct mtk_eth *eth)
@@ -4455,7 +4456,7 @@ static void mtk_hw_reset_monitor_work(struct work_struct *work)
 		goto out;
 
 	/* DMA stuck checks */
-	if (mtk_hw_check_dma_hang(eth))
+	if (mtk_hw_check_dma_hang(eth) && atomic_read(&eth->reset.force))
 		schedule_work(&eth->pending_work);
 
 out:
@@ -5617,6 +5618,7 @@ static int mtk_probe(struct platform_device *pdev)
 
 	eth->rx_dim.mode = DIM_CQ_PERIOD_MODE_START_FROM_EQE;
 	INIT_WORK(&eth->rx_dim.work, mtk_dim_rx);
+	atomic_set(&eth->reset.force, 0);
 	INIT_DELAYED_WORK(&eth->reset.monitor_work, mtk_hw_reset_monitor_work);
 
 	eth->tx_dim.mode = DIM_CQ_PERIOD_MODE_START_FROM_EQE;
diff --git a/drivers/net/ethernet/mediatek/mtk_eth_soc.h b/drivers/net/ethernet/mediatek/mtk_eth_soc.h
index f4f6ea355c52..14c3ce16f733 100644
--- a/drivers/net/ethernet/mediatek/mtk_eth_soc.h
+++ b/drivers/net/ethernet/mediatek/mtk_eth_soc.h
@@ -1479,6 +1479,7 @@ struct mtk_eth {
 
 	struct {
 		struct delayed_work monitor_work;
+		atomic_t force;
 		u32 wdidx;
 		u8 wdma_hang_count;
 		u8 qdma_hang_count;
-- 
2.30.2

